processfamily/__init__.py

```diff
@@ -519,81 +655,92 @@ def start(self, timeout=30):
-                        stderr=(subprocess.PIPE if self.ECHO_STD_ERR else open(os.devnull, 'w')) if self.CHILD_COMMS_STRATEGY else None,
-                        close_fds=self.CLOSE_FDS))
+            p = self.get_Popen_class()(cmd, **self.get_Popen_kwargs(i, close_fds=self.CLOSE_FDS))
 
             if self.CPU_AFFINITY_STRATEGY and p.poll() is None:
                 try:
                     self.set_child_affinity_mask(p.pid, i)
                 except Exception as e:
-                    logger.error("Unable to set affinity for process %d: %s", p.pid, e)
-            self.child_processes.append(_ChildProcessProxy(p, self.ECHO_STD_ERR, i, self))
+                    logger.error("Unable to set affinity for %s process %d: %s", self.get_child_name(i), p.pid, e)
+            self.child_processes.append(self.CHILD_COMMS_STRATEGY(p, self.ECHO_STD_ERR, i, self))
 
         if sys.platform.startswith('win') and self.WIN_PASS_HANDLES_OVER_COMMANDLINE:
             logger.debug("Waiting for child stream duplication events")
             for c in self.child_processes:
                 c._process_instance.wait_for_child_stream_duplication_event(timeout=timeout-(time.time()-s)-3)
 
-        if self.CHILD_COMMS_STRATEGY == CHILD_COMMS_STRATEGY_PROCESSFAMILY_RPC_PROTOCOL:
-            logger.debug("Waiting for child start events")
-            responses = self.send_command_to_all("wait_for_start", timeout=timeout-(time.time()-s))
-            for i, r in enumerate(responses):
-                if r is None:
-                    if self.child_processes[i]._process_instance.poll() is None:
-                        logger.error(
-                            "Timed out waiting for %s (PID %d) to complete initialisation",
-                            self.get_child_name(i),
-                            self.child_processes[i]._process_instance.pid)
-                    else:
-                        logger.error(
-                            "%s terminated with response code %d before completing initialisation",
-                            self.get_child_name(i),
-                            self.child_processes[i]._process_instance.poll())
-        logger.info("All child processes initialised")
+        self.wait_for_start(timeout - (time.time()-s))
+        logger.info("All child processes initialised with strategy %s", self.CHILD_COMMS_STRATEGY.__name__)
+
+    def wait_for_start(self, timeout):
+        """Waits (a maximum of timeout) until all children of process_family have started"""
+        end_time = time.time() + timeout
+        command_processes = []
+        try:
+            for child_process in self.child_processes:
+                command_processes.append(child_process.monitor_child_startup(end_time))
+            for c in command_processes:
+                # ping the process
+                c.next()
+            # results
+            return [c.next() for c in command_processes]
```

Comment:
It depends on the strategy. It is meant to be an array of results; most strategies give `None` results, but the `ProcessFamilyRPCProtocolStrategy` only gives `None` if startup failed, but otherwise puts an actual response received and parsed in a JSON response from the child